File: backend/app/schemas/link_in_bio.py
from datetime import datetime

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# BioLink schemas
# ---------------------------------------------------------------------------

class BioLinkCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    url: str = Field(..., min_length=1)
    icon: str | None = Field(None, max_length=50)
    thumbnail_url: str | None = None


class BioLinkUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=255)
    url: str | None = Field(None, min_length=1)
    icon: str | None = None
    thumbnail_url: str | None = None
    is_active: bool | None = None


class BioLinkResponse(BaseModel):
    id: str
    bio_page_id: str
    title: str
    url: str
    icon: str | None = None
    thumbnail_url: str | None = None
    position: int
    is_active: bool
    click_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# BioPage schemas
# ---------------------------------------------------------------------------

class BioPageCreate(BaseModel):
    slug: str = Field(..., min_length=1, max_length=100, pattern=r"^[a-zA-Z0-9_-]+$")
    title: str = Field(..., min_length=1, max_length=255)
    bio: str | None = None
    avatar_url: str | None = None
    theme: str = Field("default", max_length=50)
    bg_color: str = Field("#ffffff", max_length=20)
    text_color: str = Field("#000000", max_length=20)
    button_style: str = Field("rounded", max_length=20)


class BioPageUpdate(BaseModel):
    slug: str | None = Field(None, min_length=1, max_length=100, pattern=r"^[a-zA-Z0-9_-]+$")
    title: str | None = Field(None, min_length=1, max_length=255)
    bio: str | None = None
    avatar_url: str | None = None
    theme: str | None = Field(None, max_length=50)
    bg_color: str | None = Field(None, max_length=20)
    text_color: str | None = Field(None, max_length=20)
    button_style: str | None = Field(None, max_length=20)
    is_published: bool | None = None


class BioPageResponse(BaseModel):
    id: str
    user_id: str
    slug: str
    title: str
    bio: str | None = None
    avatar_url: str | None = None
    theme: str
    bg_color: str
    text_color: str
    button_style: str
    is_published: bool
    total_views: int
    created_at: datetime
    updated_at: datetime
    links: list[BioLinkResponse] = []

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Public view (no analytics fields exposed)
# ---------------------------------------------------------------------------

class PublicBioLinkResponse(BaseModel):
    id: str
    title: str
    url: str
    icon: str | None = None
    thumbnail_url: str | None = None
    position: int
    is_active: bool

    model_config = {"from_attributes": True}


class PublicBioPageResponse(BaseModel):
    id: str
    slug: str
    title: str
    bio: str | None = None
    avatar_url: str | None = None
    theme: str
    bg_color: str
    text_color: str
    button_style: str
    links: list[PublicBioLinkResponse] = []

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Reorder & Analytics
# ---------------------------------------------------------------------------

class ReorderLinksRequest(BaseModel):
    link_ids: list[str]


class ClicksByDay(BaseModel):
    date: str
    count: int


class ClickAnalyticsResponse(BaseModel):
    link_id: str
    link_title: str
    total_clicks: int
    clicks_by_day: list[ClicksByDay] = []

File: frontend/src/api/linkInBio.ts
import { api } from './client';

export interface BioLink {
  id: string;
  title: string;
  url: string;
  icon: string | null;
  thumbnail_url: string | null;
  position: number;
  is_active: boolean;
  click_count: number;
}

export interface BioPage {
  id: string;
  slug: string;
  title: string;
  bio: string | null;
  avatar_url: string | null;
  theme: string;
  bg_color: string;
  text_color: string;
  button_style: string;
  is_published: boolean;
  total_views: number;
  links: BioLink[];
}

export const listBioPages = () =>
  api.get<BioPage[]>('/bio/bio-pages').then(r => r.data);

export const createBioPage = (data: { slug: string; title: string; bio?: string; theme?: string; bg_color?: string; text_color?: string; button_style?: string }) =>
  api.post<BioPage>('/bio/bio-pages', data).then(r => r.data);

export const getBioPage = (pageId: string) =>
  api.get<BioPage>(`/bio/bio-pages/${pageId}`).then(r => r.data);

export const updateBioPage = (pageId: string, data: Partial<BioPage>) =>
  api.put<BioPage>(`/bio/bio-pages/${pageId}`, data).then(r => r.data);

export const deleteBioPage = (pageId: string) =>
  api.delete(`/bio/bio-pages/${pageId}`);

export const addBioLink = (pageId: string, data: { title: string; url: string; icon?: string }) =>
  api.post<BioLink>(`/bio/bio-pages/${pageId}/links`, data).then(r => r.data);

export const updateBioLink = (pageId: string, linkId: string, data: Partial<BioLink>) =>
  api.put<BioLink>(`/bio/bio-pages/${pageId}/links/${linkId}`, data).then(r => r.data);

export const deleteBioLink = (pageId: string, linkId: string) =>
  api.delete(`/bio/bio-pages/${pageId}/links/${linkId}`);

export const reorderBioLinks = (pageId: string, linkIds: string[]) =>
  api.put(`/bio/bio-pages/${pageId}/links/reorder`, { link_ids: linkIds });

export const getBioAnalytics = (pageId: string, days?: number) =>
  api.get(`/bio/bio-pages/${pageId}/analytics`, { params: { days } }).then(r => r.data);
